Add files via upload (#279)
Oneflow-Inc/oneflow-documentation: en/docs/extended_topics/implement_data_loader.md

As described in [Data Input](../basics_topics/data_input.md), OneFlow supports two ways to load data: one is directly use Numpy data, the other one is use DataLoader and some relative operators. 

> Under the large industrial scene, data loading can easily become the bottleneck through the training process. In other frameworks, the data loading pipeline often exists as a separate module that needs to be adapted for different scenarios which is not really versatile. In OneFlow, the DataLoader and other preprocessing operators are on same level as other common operators and can be accelerated in the same way as other operators which can easily solving the issue of large-scale data loading.

Reviewer: ```suggestion
Under the large industrial scene, data loading can easily become the bottleneck through the training process. In other frameworks, the data loading pipeline often exists as a separate module that needs to be adapted for different scenarios which is not really versatile. In OneFlow, the DataLoader and other preprocessing operators are on same level as other common operators and can be accelerated in the same way as other operators which can easily solve the issue of large-scale data loading.
```

This article will take Mini Dataloader as an example to explain the key points of implementing customized DataLoader. 

# The Composition of Dataloader

A complete Dataloader generally includes two types of Op: 

> - Data Reader: It is responsible for taking data from the file and loading it into the input flow in memory. Then ultimately setting the data to the output of the Op. It can be divided into Loader and Parser where Loader is responsible for reading raw data from the file system and Parser is responsible for organizing the raw data into the output of a Data Reader Op.

Reviewer: ```suggestion
- Data Reader: It is responsible for taking data from the file and loading it into the input stream in memory. Then setting the data to the output of the Op. It can be divided into Loader and Parser where Loader is responsible for reading raw data from the file system and Parser is responsible for organizing the raw data into the output of a Data Reader Op.
```

};
```

> As you can see, in addition to our own `MiniDataset` from `DataSet`, OneFlow also has a built-in `XXXDataSet` called **Modifier**.

Reviewer: ```suggestion
As you can see, in addition to our own `MiniDataset` from `DataSet`, OneFlow also has a built-in `XXXDataSet` called **Decorator**.
```

装饰器这种专有名词，不能随意翻译，请检查其它地方是否有错误翻译装饰器的地方。

> DataSet can add additional functionality to the existing `DataSet`, such as the above `BatchDataset` for batch reading. DataSet modifiers are located in the [user/data directory](https://github.com/Oneflow-Inc/oneflow/tree/master/oneflow/user/data) and common decorators are:

Reviewer: ```suggestion
Decorator can add functionality to the existing `DataSet`, such as the above `BatchDataset` for batch reading. DataSet decorator are located in the [user/data directory](https://github.com/Oneflow-Inc/oneflow/tree/master/oneflow/user/data) and common decorators are:
```

- BatchDataset：For batch reading data.
- RandomShuffleDataset：Used to randomize the order of data
- GroupBatchDataset：For a more customized grouping of batches and data with the same group id are placed in the same batch. More information can be found [here](https://github.com/Oneflow-Inc/oneflow/blob/267860bca25146f5053e8c878fe51231c1ced4ad/oneflow/user/data/coco_data_reader.cpp#L40). 
- DistributedTrainingDataset：For the distributed case where the data within an epoch is distributed equally among different nodes. More information can be found [here](https://github.com/Oneflow-Inc/oneflow/blob/267860bca25146f5053e8c878fe51231c1ced4ad/oneflow/user/data/distributed_training_dataset.h#L55)

After everything is done, finally call the `StartLoadThread`. As the name implies, starting the loading thread in `StartLoadThread` and it will eventually trigger to override the `MiniDataset::Next`.

> The above structure of `MiniDataReader` can be used as a template with no special requirements. There is no need to modify it during the implementation of the custom DataLoader.

Reviewer: ```suggestion
The above structure of `MiniDataReader` can be used as a template. With no special requirements, there is no need to modify it during the implementation of the custom DataLoader.
```
